## Email Spam Detection using Machine Learning

In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [10]:
df= pd.read_csv('mail_data.csv')
df.head(6)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
5     spam  FreeMsg Hey there darling it's been 3 week's n...

In [12]:
data = df.where((pd.notnull(df)),'')
data.head(5)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Category  5572 non-null   object
 1   Message   5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [15]:
data.shape

(5572, 2)

In [20]:
data.loc[data['Category'] == 'spam', 'Category',] = 0
data.loc[data['Category'] == 'ham', 'Category',] = 1

In [21]:
X= data['Message']
Y= data['Category']

In [22]:
print(X)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                 Will ü b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: Message, Length: 5572, dtype: object


In [23]:
print(Y)

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: Category, Length: 5572, dtype: object


In [27]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.33)
print(X_train.shape)
print(X_test.shape)

(3733,)
(1839,)


In [29]:
feature_extraction = TfidfVectorizer(min_df= 1, stop_words= 'english')
X_train_features = feature_extraction.fit_transform(X_train)
X_test_features= feature_extraction.transform(X_test)

Y_train = Y_train.astype('int')
Y_test = Y_test.astype('int')

In [30]:
print(X_train)

4586    U have a Secret Admirer who is looking 2 make ...
4225    Double eviction this week - Spiral and Michael...
682                         Ü log off 4 wat. It's sdryb8i
4655                     Hope you are having a great day.
5522                  Thats cool. I want to please you...
                              ...                        
1222                               Prakesh is there know.
3284                    Hey tmr maybe can meet you at yck
1627                            Dear how you. Are you ok?
4368    I like dis sweater fr mango but no more my siz...
3590    I've sent my wife your text. After we buy them...
Name: Message, Length: 3733, dtype: object


In [31]:
print(X_train_features)

  (0, 133)	0.3665625835379654
  (0, 5723)	0.3491426583025315
  (0, 194)	0.3491426583025315
  (0, 5585)	0.24880836167203566
  (0, 6334)	0.16957944223143684
  (0, 6022)	0.28958354053358654
  (0, 5077)	0.30700346576902043
  (0, 1733)	0.23376267705947762
  (0, 3803)	0.21506018428438528
  (0, 3694)	0.2635422359439501
  (0, 726)	0.30700346576902043
  (0, 5262)	0.2974165891107562
  (1, 5089)	0.44053434730134117
  (1, 2788)	0.2159179557414268
  (1, 3931)	0.44053434730134117
  (1, 5609)	0.44053434730134117
  (1, 6544)	0.2527161230087994
  (1, 2341)	0.44053434730134117
  (1, 2129)	0.33649955861927944
  (2, 5251)	0.7179421117822907
  (2, 6506)	0.42121255632581134
  (2, 3675)	0.5542013230975363
  (3, 1919)	0.4353229011807907
  (3, 2838)	0.49152127810292634
  (3, 2940)	0.5815753445350891
  :	:
  (3728, 3478)	0.43021652640346114
  (3729, 6739)	0.5779753409150482
  (3729, 3871)	0.4430498540011999
  (3729, 6090)	0.463510153829983
  (3729, 2988)	0.34357140364326005
  (3729, 3892)	0.36982206540274865
  

In [32]:
 model= LogisticRegression()
model.fit(X_train_features, Y_train)

LogisticRegression()

In [33]:
prediction_on_training_data = model.predict(X_train_features)
accuracy_on_training_data = accuracy_score(Y_train, prediction_on_training_data)

In [34]:
print("Acc on training data: ", accuracy_on_training_data)

Acc on training data:  0.9641039378515939


#### Summary:
    As we can see our model accuracy is 96% which is good , so our model is ready for future predictions.